ocaml/xapi/livepatch.ml

```diff
@@ -0,0 +1,279 @@
+    ; ("to_version", option_to_json lp.to_version)
+    ; ("to_release", option_to_json lp.to_release)
+    ]
+
+let of_json js =
+  try
+    let open Yojson.Basic.Util in
+    {
+      component= member "component" js |> to_string |> component_of_string
+    ; base_build_id= member "base_build_id" js |> to_string
+    ; to_version= member "to_version" js |> to_string_option
+    ; to_release= member "to_release" js |> to_string_option
+    }
+  with e ->
+    let msg = "Can't construct a livepatch from json" in
+    error "%s - %s: %s"
+      (ExnHelper.string_of_exn e)
+      msg
+      (Yojson.Basic.pretty_to_string js) ;
+    raise Api_errors.(Server_error (internal_error, [msg]))
+
+let get_latest_livepatch lps =
+  List.map (fun (_, _, t_v, t_r) -> (t_v, t_r)) lps
+  |> Rpm.get_latest_version_release
+  |> function
+  | Some (v, r) ->
+      List.find_opt (fun (_, _, t_v, t_r) -> t_v = v && t_r = r) lps
+  | None ->
+      None
+
+module BuildId = struct
+  let one_byte =
+    let open Angstrom in
+    any_char >>= fun c ->
+    let code = Char.code c in
+    return (Printf.sprintf "%0x" code)
+
+  let all_bytes = Angstrom.many one_byte
+end
+
+module KernelLivePatch = struct
+  let get_running_livepatch () =
+    let r =
+      (* This expects the kernel livepatch module name should be built as:
+       * livepatch_<base_version>__<base_release>__<to_version>__<to_releaes>
+       * The "." in <base_version>, <base_release>, <to_version>, and <to_releaes>
+       * should be replaced with "_".
+       * For example:
+       * livepatch_4_19_19__8_0_20__4_19_19__8_0_22
+       *)
+      Re.Posix.compile_pat
+        {|^[ ]*livepatch_([^ \[]+)__([^ \[]+)__([^ \[]+)__([^ \[]+).*$|}
+    in
+    Helpers.call_script !Xapi_globs.kpatch_cmd ["list"]
+    |> Astring.String.cuts ~sep:"\n"
+    |> List.fold_left
+         (fun (acc, acc_flag) line ->
+           match
+             ( acc_flag
+             , Astring.String.is_prefix ~affix:"Installed patch modules:" line
+             )
+           with
+           | true, _ ->
+               (acc, true)
+           | false, true ->
+               (acc, true)
+           | false, false ->
+               (line :: acc, false)
+         )
+         ([], false)
+    |> (fun (x, _) -> x)
+    |> List.filter_map (fun line ->
+           let replace s =
+             Astring.String.cuts ~sep:"-" s |> Astring.String.concat ~sep:"."
+           in
+           match Re.exec_opt r line with
+           | Some groups ->
+               Some
+                 ( Re.Group.get groups 1 |> replace
+                 , Re.Group.get groups 2 |> replace
+                 , Re.Group.get groups 3 |> replace
+                 , Re.Group.get groups 4 |> replace
+                 )
+           | None ->
+               None
+       )
+    |> get_latest_livepatch
+
+  let align x =
+    (* The actual size is aligned with 4. *)
+    let open Int32 in
+    (if rem x 4l = 0l then x else add x (sub 4l (rem x 4l))) |> to_int
```

Comment:
The code is using `41` but the comment says something is 4-byte aligned. What is the relevance of `41`? What is the result of this function for a given x?

Reply:
Sorry, I misread this. The `1` is actually an `l` - a size indicator. A comment what `align 6` would do would be still helpful. Are we rounding up or down to a multiple of 4? I assume `align x` rounds x up to the next multiple of 4. I believe it could be simplified (here for integers):
```

Reply:
Yeah, this is great. Thanks.